### DISCUSSION TITLE
Audit by CS

### DISCUSSION THREAD

**Discussion by brozorec** [Fujicracy/fuji-v2#83] [General]
Our smart contracts will be audited by Composable Security in two tranches. The first tranch of contracts should be ready by **Oct 26th.** 

## 1st Tranch

1. We have to define the scope of this first tranch and we need to select contracts that:
- contain the core functionalities of the system
- are stable, i.e. they aren't supposed to change a lot in the future.

This is my proposal:

| candidates | readiness | what's left |
|----------------|---------------|---------------|
| Chief | 60% | acl and roles; store vaults list |
| FujiOracle | 90% | tests |
| BaseVault | 85% | tests |
| BorrowingVault | 70% | acl; tests |
| BorrowingVaultFactory | 90% | tests |
| YieldVault | 70% | acl; tests |
| YieldVaultFactory | 90% | tests |
| AaveV2 (Ethereum, Polygon) | 95% | clone |
| AaveV3 (Polygon, Optimism, Arbitrum) | 90% | clone |
| Compound V2 | 75% | copy from alpha |
| Compound V3 | 100% | none |

2. There is one topic touching on almost all of the candidate contracts which #78  
 
3. We need to write more tests cases (we may reorganize "test" directory to split it into tests with mocks and tests on forks)

4. We have to create some diagrams showing how different elements of the system interact with each other

5. We have to review all code comments and add more wherever needed

## 2nd tranch: TODOs

**1. BaseRouter** (BB)
- [x] #103 
- refactor to include `DepositETH`, `BorrowETH`, `PaybackETH` and `WithdrawETH`
refs: 
https://github.com/aave/protocol-v2/blob/master/contracts/misc/WETHGateway.sol 
https://github.com/compound-finance/comet

- remove PeripheryPayments
- write tests

**2. ConnextRouter** (BB)
- [x] #102 
- adapt according their new API
- write tests

**3. Flashloans** 
_Let's have one contract per flashloan provider instead of one big contract containing all providers. In "alpha" we had to redeploy every time we wanted to add a new provider + we have to keep track of their IDs, so it doesn't make a lot of sense to keep them all together._ 
- add at least one Flasher for each chain

**4. Providers** (PV)
- what other providers we want to add?

- [x] **5. AddrMapper** (DC)
_"AddrMapper" is used in some provider contracts as a helper contract to map ERC20s to their IBT equivalents for each provider. We need to deploy a new "AddrMapper" for every new provider contract that requires such mapping. We need to consider having only one "AddrMapper" for this purpose as this will greatly simplify management._

**6. LiquidationManager** - a new contract that manages a list of allowed liquidators 
- specs
- code
- tests

**7. RebalanceManager** - a new contract that manages a list of allowed rebalancers and handles flashloan logic (DC, BB)
 - [x] #104 

**8. HarverstManager** - a new contract that manages a list of allowed harvesters and handles specificities for each protocol.
- specs
- code
- tests

cc @DaigaroCota @iafhurtado

**Comment by 0xdcota** [comment]
Thanks for putting all these togethers @brozorec. I agree, and based on the information I see the following "urgent" Issues that should be opened:
@iafhurtado please create:

1. `Create Tests missing in Tranch1 functions`: Refer to the table on the contracts that require/are missing testing. As discussed, this is probably were Pedro will jump in.  
  
2. `Clone Providers and versions in intended chains`:  Refer to the table to see  the appropriate Provider contracts and the versions that need to be in all the chains indicated.   

3.  `Complete deploy scripts for all intended chains`: this is a secondary task result of the provider creation task before,  but good to keep track of it in a separate ticket.

**Comment by iafhurtado** [comment]
Done they are:

- [ ] #87 
- [ ] #88 
- [ ] #89
